Handling of Fx variables as regular variables for CMIP5 and CMIP6 WITH CMOR checks

valeriupredoi opened the pull request:
should address comments and concerns from #901 and standardize fx files finding for CMIP6

zklaus commented:
I suggest to rename `fxdir` to `mip` and `fx_var` to `short_name`. Then we can use anchors to get the paths directly from the data configuration.

valeriupredoi commented:
was doing it while you were writing your comment, dude :grin:

valeriupredoi commented:
OK so I've implemented the fx variable retrieval as a regular CMOR variable with its files and all other properties that a regular variable has; this implies that fx var files will have to be listed in the `variables` section of the diagnostic with the `default` preprocessor eg:
```
    variables:
      tas:
        preprocessor: pp_rad
        mip: Amon
        grid: gr
        fx_files: ['areacello']
      areacello:
        fxvar: True
        grid: gn
        mip: Ofx
```
Note that:
 - the mandatory identification of the variable to be an fx variable is done via the `fxvar: True` flag
 - grids need to be specified since lo and behold, they sometimes differ from the other variables' ones (if it's ocean then they are 'gn' and if all other vars are ocean then there is no need to specify grid in the fx var)

With this, the preproc output dir will contain, on top of all the other diag vars, CMOR-checked files for the needed fx vars, which is good. What's bad is that some of these vars don't have CMOR table entries (like `areacello`) and will need custom CMOR files; they may also need fix files, dunno yet.

One thing I couldn't manage for the life of me is to assign these preproc-ed (CMOR checked) fx files to the `variable['fx_files']` instead of (currently) just retrieving the files from the original location (which works now, but it's not what I want) - @bouweandela I needs your help here, man! What's the best strategy to allow for this since the preprocessor steps are independent and done per variable?

zklaus commented:
> OK so I've implemented the fx variable retrieval as a regular CMOR variable with its files and all other properties that a regular variable has; this implies that fx var files will have to be listed in the `variables` section of the diagnostic with the `default` preprocessor eg:
> 
> ```
>     variables:
>       tas:
>         preprocessor: pp_rad
>         mip: Amon
>         grid: gr
>         fx_files: ['areacello']
>       areacello:
>         fxvar: True
>         grid: gn
>         mip: Ofx
> ```
> 
> Note that:
> 
>  the mandatory identification of the variable to be an fx variable is done via the `fxvar: True` flag

Maybe for consistencies sake that should be `fx_var`.

>  grids need to be specified since lo and behold, they sometimes differ from the other variables' ones (if it's ocean then they are 'gn' and if all other vars are ocean then there is no need to specify grid in the fx var)

The grid of a fx variable is always the same as the pertinent data variable. The example above is a bit strange in that it gives the cell areas for the ocean grid (`areacell*o*`) together with some atmospheric data. If you look at the corresponding `areacell*a*`, you will surely find the grid you expect.

Apart from that, `areacella/o` along with a few others are a bit special (see my last comment in the issue
https://github.com/ESMValGroup/ESMValTool/issues/901#issuecomment-487598899). Actually, allow me to quote myself:
> Additionally, some of these are actually `cell_measures`, so the most elegant way, imho, to deal with them would be to add the fx files to the list of input filenames before they are loaded as cubes into iris.
> Then, iris will automatically sort them out and make them available via `cube.cell_measure()`.
> Grepping the tables, this seems to apply for
>
> - `areacella` from fx
> - `areacellg` from IfxAnt and IfxGre
> - `areacello` from Ofx
> - `volcello` from either Ofx or (Omon, Oyr, Odec).
> - `areacellr` from fx

Also, if they are used in a `.nc` file, its global attributes will list them as `external_variables` in accordance with [the relevant section of the CF conventions](http://cfconventions.org/Data/cf-conventions/cf-conventions-1.7/cf-conventions.html#external-variables). So maybe we should be able to use that information automatically?

> With this, the preproc output dir will contain, on top of all the other diag vars, CMOR-checked files for the needed fx vars, which is good. What's bad is that some of these vars don't have CMOR table entries (like `areacello`) and will need custom CMOR files; they may also need fix files, dunno yet.

Hm, `areacello` is in `Ofx`. Which ones don't have entries?

valeriupredoi commented:
hey K-man @zklaus good stuff, cheers! also my previous comment is getting obsolete now since I am gradually plugging in all the changes needed to make things work properly. Stay tuned :radio:

valeriupredoi commented:
ok guys, so here it is: now we can do the following:

 - specify orography fx variables for both CMIP5 and CMIP6 in the recipe with as little as eg 
```
    variables:
      tas:
        preprocessor: simple_preproc
        mip: Amon
      sftlf:
        fxvar: True
      sftof:
        fxvar: True
```
 - the code will retrieve the data from the DB, apply cmor checks and fixes, and will run the needed preprocessor steps that use the fx files with the cmor checked and fixed output files;
 - tasking is now done with priority ie if fx variables are present in a diagnostic, those will be processed first (preprocessor=default, basically just find data, load, fix and cmor check/fix; of course they can be run through a more complex preproc but I fail to see why) then the other variables will be run; this we we are certain that the output fx files (now stored in the `preproc/` dir like any variable) will be there when needed by any given preprocessing or diag step; there is a bit of a wait on the main meat of preproc but these files are small and they are fast at preproccing;
 - I tested both with CMIP5 and 6 and all went well for me, there is a downside though - in cases like mask_landsea one needs to actually provide the fx variables in the recipe now (sftlf, stflof etc) becasue I could not work a way to add them to the list of variables without actually decalring them in the recipe - maybe @bouweandela has an idea?

Also @bouweandela  - with this ine, the rather bulky implementation in #1009 is now obsolete; you will see now that `_recipe.py` does what its job is - just sets up a workflow of tasks and executes nothing internally :beer: 

Another thing - now that the fx variables/files are available as regular variable files - how do we get about to including the cell measures in the actual variable? I gots no idea

valeriupredoi commented:
**Pending approval**

 - close #1009 
 - close #901 
 - close #547

zklaus commented:
Good stuff, @valeriupredoi !

I wonder: Would it be possible/reasonable to forgo the `fxvar` in favor of the normal `mip`? Afterall, we know that fx files are exactly those that have `fx` in the name of their table. Or, even more explicit, we can give a list of tables of fixed files.

valeriupredoi commented:
OK guys a few (and final changes): as mentioned above the fx variables can now be standard diagnostic variables; changes of the last commit: it is now enough to specify a list of fx variables eg `fx_files: ['sftlf', 'sftof']` in the variable and the code will pick up the fx files, CMOR check and fix them and then apply whatever mask is needed in the preprocessor; NOTE that if one only specifies the mask in the preproc and no fx settings in variable OR separate fx variables as stand alone diagnostic variables, the masking will fail. I need to update the doeumentation but after people test and also give me coding feedback.

@zklaus in essence this PR does exactly what you are saying - you can now just list whatever fx variables you need to include as diagnostic variables; I may drop the `fxflag` completely from the recipe, yes - the fx var needs an internal identification flag though because the DRS is different from a regular variable, and also it runs a very stripped-down preprocessor profile (eg no time selection, no regridding etc)

zklaus commented:
Nice! In what way is the drs different? looking at your `config-developer.yml`, the dir part seems to be identical and the file part only differs in that the general files have `_*.nc` and the fx files have `*.nc` in the end. If that really is the only difference we might consider simply removing the extra `_`?

valeriupredoi commented:
OK last commit promise :grin: 

So now we have the following features: both for CMIP5 and CMIP6:

 - specifying an `fx_files: [list of fx variables short names]` in each variable will: find the correct fx files from remote; apply cmor checks and fixes; write the cmor checked and fixed files to `/preproc/` under dirs bearing the names of the fx variables; link the **outputted** cmor checked and fixed files in the `fx_files: {}` dictionary in the variable and use the same **outputted* cmor checked and fixed files for landsea/ice masking or derive; pretty fucken ninja is you ask me :grin: 
 - CAVEAT: if one wants to specify eg `mask_lansea` in the preprocessor, one also needs to specify eg `fx_files: ['sftlf', 'sftof']` for each variable with that preprocessor;
 - is is also possible to list the fx variables as individual variables in the recipe, but there is no point to do so and it's also bad practice since the recipe checks will throw out a no existing ancestors for those variables error;

@zklaus - for CMIP6 fx variables the mip differs from the parent variable; this shouldn't be a problem in the future since I have a function that maps fx variable short name to mip from the config developer; if we decide to use the same data finder function for the fx variables, it'll only need a small tweak, but I reckon this PR is huge enough to leave it as is now; apart from that all other variable properties apply to the fx variables as well ie you can see that `Recipe._initialize_variables()` doesn't care if the variables list contains fx variables or not, they are thrown in together with the parent variables

@bouweandela have a look, man, and tell me what you thinks :beer: 

@mattiarighi it'd be great if you could tested, man :beer:

mattiarighi commented:
Thanks! I will test after @bouweandela's approval, as usual.

One question: does this require adjustments to the existing recipes?
If yes, we might need more extensive testing.

valeriupredoi commented:
if @bouweandela has a stroke of genius and tells me how to add the fx_files list in the parent variable member without having to specify it explicitly in the recipe when using a mask_landsea or mask_landice preproc, then no recipes will need changing; if not, then the recipes that use either mask_landsea or mask_landice will have to get the fx_files list plugged in the recipe under the variables that use the mask preproc but that's it, nothing else

valeriupredoi commented:
circleCI is drunk today, tests pass fine locally (on two different machines), dunno what's up with it

bouweandela commented:
Yes, it should be done with ancestors instead.

valeriupredoi commented:
something's very wrong with the comments time labels - anyways, I am now using ancestors and have reverted the changes to `_task.py` cheers @bouweandela :beer:

valeriupredoi commented:
looks like @bouweandela and myself could post comments in the future :back: 

In any case, I am finally happy with this PR - it's big, and I have to apologize for this, but it does what it says on the tin: grabs fx variables from parent diagnostic variables and treats them as normal variables with all the bells and whistles of our cmor checks and fixes; I have now reverted all the changes to `_task.py` and all the prerequisite analysis steps are done via ancestors; also note that the cmorized fx files now live in the parent variable preproc dir so that they are available for further concatenation and use in the diagnostic; there are a few bits of restructuring in `_recipe.py` but things should be clear and I tried to avoid code clutter as much as possible. Good luck @bouweandela - we can do the review together on the phone if it'd be easier for you :beer: I have tested thoroughly with CMIP5 datas, limited testing with CMIP6 since only a handful of fx variables are available just yet

valeriupredoi commented:
take your time @bouweandela - the only change I made to `_task.py` was to allow for preferential tasks to be completed first ie the fx variable processing needs to be done first so that the regular variable can use the output - but if you can do that with ancestors then we wouldn't need to change `_task.py` (I didn't know how to work that out, the ancestors, that is)

zklaus commented:
Cheers @valeriupredoi  :beer: !

Just to clarify:
Basically there are three ways that fx variables can find their way into an analysis

1. as `cell_measure`s
2. as additional info for the preprocessor
3. as variables of their own right (say, if you want to write a diagnostic that compares the land areas of different models)

Could you help me understand how these cases are handled?
For example, for 2., I have a feeling what could be considered a parent variable, but in general, fx variables are applicable to many different variables and can even stand on their own, right?

zklaus commented on esmvaltool/config-developer.yml:
```diff
@@ -25,6 +25,12 @@ CMIP6:
+  input_fx_dir:
+    default: '/'
+    BADC: '[institute]/[dataset]/[exp]/[ensemble]/[mip]/[short_name]/[grid]/[latestversion]'
+  input_fx_file: '[short_name]_[mip]_[dataset]_[exp]_[ensemble]_[grid].nc'
```
Do we really need `input_fx_dir` and `input_fx_file`? It seems to be covered by `input_dir` and `input_file`. The only necessary change is to remove the extra `_` in front of the `*` in `input_file`.

valeriupredoi replied:
yes, exactly, but will keep this until we remove (if we do that) the extra string template for regular variable files; note that the file finder function will still need to be a special one since mip for CMIP5 and CMIP6 needs to be handled separately for fx variables

zklaus replied:
> yes, exactly, but will keep this until we remove (if we do that) the extra string template for regular variable files

I don't know what you mean with "the extra string template for regular variable files". Is it just `input_file`?

> note that the file finder function will still need to be a special one since mip for CMIP5 and CMIP6 needs to be handled separately for fx variables

Hm. Just looking at `config-developer.yml`, this does *not* seem to be the case: Also in CMIP5 these variables have `[mip]` (`fx`), `[frequency]`(`fx`), and `[ensemble]` (`r0i0p0`), bringing the templates in exact agreement with those for all the other normal variables.

valeriupredoi replied:
have a look at `_config,replace_mip_fx(fx_file)` and `_config._get_cmip6_fx_mip(fx_var_name)` - there is just not enough information encoded in the `input_file` and `input_dir` from config-developer to be able to retrieve the correct fx files using just those, at the moment

zklaus replied:
Maybe I don't understand, but if you look at the lines above they are literally identical, so how does this add any information?

valeriupredoi commented:
@zklaus man: for the `cell_measure` types I believe they are already part of the netCDF file of the parent variable and at load point the cubeList will contain the parent variable and the cell_measure cubes; there is, however, a problem there that we need to address - the tool will attempt to concatenate those cubes and will fail, but that's not a problem concerning this PR, that needs to be addressed in a different PR; for the additional info for the preprocessor - it's this PR: so say variable tas is specified this way:
```
tas:
  preprocessor: some_pp
  mip: Amon
  fx_files: [sftlf, sftof]
```
the code will now grab the `fx_files` list and convert the fx variables strings into full blown CMOR variables, get their files, apply a basic 9default) preprocessor on them (that will not include time gating, but all else in the defult preproc like load, fix, cmor check and fix, save and cleanup) and write the output files in the `preproc/diag_name/tas` directory and also populate the `tas` metadata file with the `fx_files` dictionary; it will also use those output files to apply any mask that uses either `sftlf` or `sftof` and is asked for in the preprocessor eg `mask_landsea`
For the stand alone fx variable asked for in the variables list eg:
```
sftgif
  fxvar: True
```
This is also possible via this PR, and the variable will be treated as any other CMOR variable with the only difference that the file finder will use the slightly different fx file finder function (until, as you point out above, we decide to merge those into one file finder function). Note that since this is a stand alone variable now, the preprocessor output for it will be written in a standard location `preproc/diag_name/sftgif` but the preprocessor profile for it will still be the default one bar time gating since we don't want to apply any regridding or any other special situation

zklaus commented:
> for the `cell_measure` types I believe they are already part of the netCDF file of the parent variable...

Unfortunately, this is not true. Usually, they are named as [`external_variables`](http://cfconventions.org/Data/cf-conventions/cf-conventions-1.7/cf-conventions.html#external-variables) and as such are not in the netCDF file.

> ...and at load point the cubeList will contain the parent variable and the cell_measure cubes; there is, however, a problem there that we need to address - the tool will attempt to concatenate those cubes and will fail, ...

I'm not sure if ESMValTool changes this, but iris should deal with it just fine[^1]. If the cell_measure is available at load time, even if it is in a different file, it will be incorporated correctly.

> ...but that's not a problem concerning this PR, that needs to be addressed in a different PR.

Fair enough. I'll add an issue then.

> For the additional info for the preprocessor - it's this PR: so say variable tas is specified this way:
> 
> ```
> tas:
>   preprocessor: some_pp
>   mip: Amon
>   fx_files: [sftlf, sftof]
> ```
> 
> the code will now grab the `fx_files` list and convert the fx variables strings into full blown CMOR variables, get their files, apply a basic 9default) preprocessor on them (that will not include time gating, but all else in the default preproc like load, fix, cmor check and fix, save and cleanup) and write the output files in the `preproc/diag_name/tas` directory and also populate the `tas` metadata file with the `fx_files` dictionary; it will also use those output files to apply any mask that uses either `sftlf` or `sftof` and is asked for in the preprocessor eg `mask_landsea`

Nice! Where do you specify the `[mip]` for the fx variables?

> For the stand alone fx variable asked for in the variables list eg:
> 
> ```
> sftgif
>   fxvar: True
> ```
> 
> This is also possible via this PR, and the variable will be treated as any other CMOR variable with the only difference that the file finder will use the slightly different fx file finder function (until, as you point out above, we decide to merge those into one file finder function). Note that since this is a stand alone variable now, the preprocessor output for it will be written in a standard location `preproc/diag_name/sftgif` but the preprocessor profile for it will still be the default one bar time gating since we don't want to apply any regridding or any other special situation

Nice. I assume you forgot the
```
  mip: fx
```
?


[^1]: I wanted to reconfirm this, but unfortunately ESGF is very unresponsive at the moment and doesn't permit me to download the test files.

valeriupredoi commented:
K-man, so in brief:
 - first case (`cell_measures`) - we need to talk in detail about that, let's issue it separately;
 - second case: the mip is fx for CMIP5 and corrected for via config-developer for CMIP6 and this is exactly why we (at the moment) need a special `get_input_fx_filelist` rather tha using the regular variables `get_input_filelist` can not be used - the mip adjustment is done automatically ie the user needs to specify only the `fx_files: list of short names` and then all else is done internally
 - third case: if the `fxvar` flag is raised then all is done as per the second case; of course, one can not raise the `fxvar` in recipe and then they will have to add `mip` and the fx variable will be no different than a regular variable, but I haven't properly tested that just yet, I will do it now

valeriupredoi commented:
so for those external variables, now that `fx_files: [external fx vars list]` in a parent variable works nice and automated (and will always output fixed files **before** the parent variable is loaded and analyzed), we can probably add a callback at the load point of the parent variable to add the external fx files to it eh?

zklaus commented:
> first case (`cell_measures`) - we need to talk in detail about that, let's issue it separately;
:+1:  Ack.

As for the rest, I guess I just don't get why we don't specify the mip with the fx vars. In CMIP every variable lives in one or more tables and it is necessary to specify the table to really be clear about the variable (look at `uas` which lives in no less than 7 tables). These variables are no different.

It seems to me that we lift the perfectly reasonable requirement to fully specify the required data including the table, at the cost of introducing a lot of special handling and hassle for us.

> second case: the mip is fx for CMIP5 and corrected for via config-developer for CMIP6 and this is exactly why we (at the moment) need a special `get_input_fx_filelist` rather tha using the regular variables `get_input_filelist` can not be used - the mip adjustment is done automatically ie the user needs to specify only the `fx_files: list of short names` and then all else is done internally

The general point above notwithstanding, I don't doubt that some extra python code is necessary. That doesn't mean we need to introduce redundancy in `config-developer.yml`.

> third case: if the `fxvar` flag is raised then all is done as per the second case; of course, one can not raise the `fxvar` in recipe and then they will have to add `mip` and the fx variable will be no different than a regular variable, but I haven't properly tested that just yet, I will do it now

Nice. This seems to me to be the normal way for this case.

valeriupredoi commented:
@zklaus and all, so here's the framework, and best way is to have an example:
```
preprocessors:
  with_mask:
    seasonal_mean:
    mask_landsea:
        mask_out: land
    regrid:
      target_grid: 1x1
      scheme: linear
  without_mask:
    seasonal_mean:
    regrid:
      target_grid: 1x1
      scheme: linear

diagnostics:
  the_best_diagnostic:
    variables:
      tas:
        preprocessor: with_mask
        mip: Amon
        fx_files: ['sftlf', 'sftof']
      pr:
        preprocessor: with_mask
        mip: Amon
        fx_files: ['sftlf', 'sftof', 'sftgif']
      sftlf:
        is_fx_variable: True
        mip: fx
      sftof:
        is_fx_variable: True
        mip: fx
      ta:
        preprocessor: without_mask
        mip: Amon
      rlns:
        preprocessor: with_mask
        derive: True
        force_derivation: False
        mip: Amon
        fx_files: ['sftlf', 'sftof']
```

 - here, `sftlf` and `sftof` as standard variables are treated and analyzed as standard variables with a `default` preprocessor that will not execute time gating;
 - the `fx_files: ['sftlf', 'sftof']` variable attribute will trigger analysis of each of those fx variables in the list as CMOR variables (checks, fixes and all); output will be in each parent variable directory eg:
```
(esmvaltool_new) [valeriu@jasmin-sci2 esmvaltool_var_test]$ ls -la FXcmip5/recipe_dask_20190509_120815/preproc/the_best_diagnostic/pr/
total 51712
drwxr-xr-x 1 valeriu users        0 May  9 13:09 .
drwxr-xr-x 1 valeriu users        0 May  9 13:09 ..
-rw-r--r-- 1 valeriu users 22337152 May  9 13:09 CMIP5_MPI-ESM-LR_Amon_historical_r1i1p1_pr_1990-2000.nc
-rw-r--r-- 1 valeriu users    98865 May  9 13:08 CMIP5_MPI-ESM-LR_fx_historical_r1i1p1_sftgif_1990-2000.nc
-rw-r--r-- 1 valeriu users    98810 May  9 13:08 CMIP5_MPI-ESM-LR_fx_historical_r1i1p1_sftlf_1990-2000.nc
-rw-r--r-- 1 valeriu users  2494538 May  9 13:08 CMIP5_MPI-ESM-LR_fx_historical_r1i1p1_sftof_1990-2000.nc
-rw-r--r-- 1 valeriu users 22337192 May  9 13:09 CMIP5_NorESM1-M_Amon_historical_r1i1p1_pr_1990-2000.nc
-rw-r--r-- 1 valeriu users    79270 May  9 13:08 CMIP5_NorESM1-M_fx_historical_r1i1p1_sftgif_1990-2000.nc
-rw-r--r-- 1 valeriu users    78251 May  9 13:08 CMIP5_NorESM1-M_fx_historical_r1i1p1_sftlf_1990-2000.nc
-rw-r--r-- 1 valeriu users  5424027 May  9 13:08 CMIP5_NorESM1-M_fx_historical_r1i1p1_sftof_1990-2000.nc
-rw-r--r-- 1 valeriu users     2585 May  9 13:09 metadata.yml
```
 - the `fx_files: ['sftlf', 'sftof']` in each variable will trigger inclusion of `fx_files: ['sftlf', 'sftof']` in each variable metadata for use in the diagnostic; also those files will be used for `mask_landsea` for `preprocessor: with_mask`;
 - same applies for derived variables eg:
```
(esmvaltool_new) [valeriu@jasmin-sci2 esmvaltool_var_test]$ ls -la FXcmip5/recipe_dask_20190509_120815/preproc/the_best_diagnostic/rlns/
total 51528
drwxr-xr-x 1 valeriu users        0 May  9 13:10 .
drwxr-xr-x 1 valeriu users        0 May  9 13:09 ..
-rw-r--r-- 1 valeriu users 22332846 May  9 13:10 CMIP5_MPI-ESM-LR_Amon_historical_r1i1p1_rlns_1990-2000.nc
-rw-r--r-- 1 valeriu users    98810 May  9 13:08 CMIP5_MPI-ESM-LR_fx_historical_r1i1p1_sftlf_1990-2000.nc
-rw-r--r-- 1 valeriu users  2494538 May  9 13:08 CMIP5_MPI-ESM-LR_fx_historical_r1i1p1_sftof_1990-2000.nc
-rw-r--r-- 1 valeriu users 22332846 May  9 13:10 CMIP5_NorESM1-M_Amon_historical_r1i1p1_rlns_1990-2000.nc
-rw-r--r-- 1 valeriu users    78251 May  9 13:08 CMIP5_NorESM1-M_fx_historical_r1i1p1_sftlf_1990-2000.nc
-rw-r--r-- 1 valeriu users  5424027 May  9 13:08 CMIP5_NorESM1-M_fx_historical_r1i1p1_sftof_1990-2000.nc
-rw-r--r-- 1 valeriu users     2357 May  9 13:10 metadata.yml
```
I think thhis covers it fully and I am ready to call this done :beer:

zklaus commented:
It's nice to have an example, but I'd rather have answers. To recap the open questions shortly:

1. Why the redundancy in `config-developer.yml`? Can we avoid it?
2. Why leave out the table information from the recipe and introduce guess work?

valeriupredoi commented:
>>> Why leave out the table information from the recipe and introduce guess work?

what do you mean? Now it needs the mip table from the recipe for an fx variable that is a stand alone recipe variable

>>> Why the redundancy in config-developer.yml? Can we avoid it?

that is still needed for the  `fx_files: ['sftlf', 'sftof']` cases; don't be bothered too much about that, a user doesn't know anything about that; as long as things are easier for the user, it'f fine for us to have a couple more lines in a dev config file

zklaus commented:
> > Why leave out the table information from the recipe and introduce guess work?

> what do you mean? Now it needs the mip table from the recipe for an fx variable that is a stand alone recipe variable

Yes, but what about the other case?
 
> > Why the redundancy in config-developer.yml? Can we avoid it?

> that is still needed for the `fx_files: ['sftlf', 'sftof']` cases;

You keep repeating that, but you don't say why. There is no additional information in those lines. Mind you, I can (scarcely) imagine the need for a different `get_input_fx_filelist`, but why couldn't that use the normal information from the other line?

> don't be bothered too much about that, a user doesn't know anything about that; as long as things are easier for the user, it's fine for us to have a couple more lines in a dev config file

I agree that the user should be in the focus. I just disagree that this makes it easier for the user. It seems confusing to learn a different way to specify variables, even if that way might appear easier. It also could fuel misconceptions less experienced users might have about the CMIP data request.

valeriupredoi commented:
man, there's two totally different cases at hand here: one that the user needs to use some fx variables for some comparison or any other analysis **with them as normal variables** (that may also involve preprocessing them) - in that case the user adds them in the recipe as any other variable (with mip!) and the second independent case is when the user needs the fx variables for masking (in the preproc via mask_something and/or in the diagnostic for whatever masking purposes) - in that case the fx_files dict is built for every variable that needs it and no mips or any other details are provided in the recipe but rather the tool builds everything that is needed under the hood; also for this case there is no preprocessing of the auxiliary fx variables apart from cmor checks and fixes. Run a test with the recipe I posted above so you see the thing in practice, rather than theorizing about it :beer:

zklaus commented:
Hi V, I completely understand how it works. What I don't understand is why you outright refuse to answer the questions.

valeriupredoi commented:
@zklaus hmm let's see, so about the `fx_files: [...]` case in the variable: currently this is coded so that the user specifies just this in the variable and nothing else, hence this information is not enough to be able to use the standard variable file finder settings; we can probably adapt it though, so that it uses the standard file finder and settings in config-developer but then again it will still need some info on the mip (for CMIP6) - that can probably use the var short_name to mip mappings that we still have to include in config-developer, let me see what I can do about this. I think this answers the second question too, no? :beer:

bouweandela commented:
Hi V.,

I've been thinking about the issue of fixes to fx variables and using fx files when loading normal variables to get the cell methods etc included, and I think the best method may be to take this out of the preprocessor and use a specially constructed iris callback at cube load time. It would need some experimentation to see if it can be done, but I think it would be simpler than trying to build a very complicated task tree.

Regarding this PR, to keep it manageable, maybe it would be best if you could just focus on what the title says 'Handling of Fx variables as regular variables for CMIP5 and CMIP6', and leave the fixes applied to fx variables that are used as input for preprocessor functions for a later pull request.

To do that, I think the following two things would need to be done:
1) remove the functions for finding fx files from data_finder, because the the normal ones can be used.
2) change the recipe format from where we have
```yaml
variables:
  pr:
    mip: Amon
    fx_files: [sftlf]
  ```
to
```yaml
variables:
  ta:
    mip: Amon
  sftlf:
    mip: fx
```
if @mattiarighi agrees with this new format, of course. Automatically determining the mip etc for fx files would of course be nice, but maybe not strictly necessary at first? If you're going to do it, I would recommend using the cmor tables to find the mip matching a particular fx variable, instead of hand copying this into config-developer.yml.

If it turns out that it is too complicated to reach an agreement here, maybe it would be best to leave this issue until the workshop in June, as it is always much easier to discuss things in person than on github.

mattiarighi commented:
I must admit I've got a bit lost through this long discussion, but I'll try to answer your question.

> if @mattiarighi agrees with this new format, of course. Automatically determining the mip etc for fx files would of course be nice, but maybe not strictly necessary at first?

For CMIP5 this is not needed, since `fx` is actually the `mip`.  But it is needed for CMIP6, I think, since the directory structure is different and more complicated.

> If you're going to do it, I would recommend using the cmor tables to find the mip matching a particular fx variable, instead of hand copying this into config-developer.yml.

If I understood @zklaus' comment above right, this would be the most robust solution.

> If it turns out that it is too complicated to reach an agreement here, maybe it would be best to leave this issue until the workshop in June, as it is always much easier to discuss things in person than on github.

I agree.

zklaus commented:
Thanks for sticking with it. I do appreciate it. :beer: 

Actually I think it comes closer to answering the second question, than the first.
The first is about these two pairs of lines in `config-developer.yml`.
The upper one is from `input_dir` and the lower one from `input_fx_dir`
```
BADC: '[institute]/[dataset]/[exp]/[ensemble]/[mip]/[short_name]/[grid]/[latestversion]'
BADC: '[institute]/[dataset]/[exp]/[ensemble]/[mip]/[short_name]/[grid]/[latestversion]'
```
and from `input_file` and `input_fx_file`
```
  input_file     : '[short_name]_[mip]_[dataset]_[exp]_[ensemble]_[grid]_*.nc'
  input_fx_file: '[short_name]_[mip]_[dataset]_[exp]_[ensemble]_[grid].nc'
```
The fact that they are identical is no coincidence, but a consequence of the fact that fx variables are just normal cmip variables.

With regards to the first question, you say that currently things are coded so that the user only specifies the variable name. The tool than tries to figure out the table. If we accept this approach, the rest follows more or less as you describe it. But my point is that this approach is questionable. The user knows that a priori he has to specify always the pair (table, variable) to identify a piece of information in cmip. When hearing that he doesn't have to give the table here, they will be a bit surprised, but accept it. The real problems for us come with variables that don't live in one table only. For example, suppose you want to do the masking not with land fraction (`sftlf`), but with ice sheet fraction (`sftflf`). Suddenly you have 4 tables to choose from due to special treatment of Greenland and Antarctica. Specifying *always* (table, variable) to refer to pieces of information in cmip is clean, simple, and universal.

:beer:

zklaus commented:
> For CMIP5 this is not needed, since `fx` is actually the `mip`. But it is needed for CMIP6, I think, since the directory structure is different and more complicated.

It's true that it might not be needed, but it also doesn't hurt since, as you rightly say `fx` *is* the `mip`. In this sense, also in CMIP5 there is nothing special going on with the fx files.

mattiarighi commented:
> The fact that they are identical is no coincidence, but a consequence of the fact that fx variables are just normal cmip variables.

You are right, the filename structure is identical for `fx` and non-`fx` variables, so `input_fx_file` is redundant. I guess this is a remnant from v1.

> The user knows that a priori he has to specify always the pair (table, variable) to identify a piece of information in cmip. When hearing that he doesn't have to give the table here, they will be a bit surprised, but accept it.

But do I understand correctly, that in CMIP5 `experiment` + `fx`-short_name fully identifies the variable, but in CMIP6 additional information are needed?

zklaus commented:
> But do I understand correctly, that in CMIP5 `experiment` + `fx`-short_name fully identifies the variable, but in CMIP6 additional information are needed?

In both CMIP5 and CMIP6 many variables across all tables live in several tables, but many also live in only one table. A search through [the CMIP5 data request](https://cmip.llnl.gov/cmip5/docs/standard_output.xls) suggests that all of the variables in `fx` only live there, but it's not quite as clear cut because for several of them the comment then states

> If this field is time-dependent then save it instead as one of your Omon fields (see the Omon table)

mattiarighi commented:
> If this field is time-dependent then save it instead as one of your Omon fields (see the Omon table)

This makes sense, since `fx` are supposed to be time-invariant.
Hopefully this is still the case in CMIP6... :eyes:

valeriupredoi commented:
OK guys, cheers muchly for all the comments -- so here's what I think we should do here:

 - I will remove the special fx file finders and use the normal variable finder;
 - fx variables should get the mip and any other cmor information from the cmor tables; to my knowledge all fx variables are time-independent, so a preprocessor on them needs to have the time gating removed (currently all these are implemented);
 - (and this is the tricky bit) allow for use of cmorized fx-variable files both for masking preprocessors and for later use in the diagnostic. This breaks down like this:
    - for diagnostic use **only** is simple and if the user puts the needed fx variables in the recipe (like all other variables) them files will be avialable at the end of preprocessor;
    - for masking done in the preprocessor, the current implementation allows for it by specifying `fx_files : [...]` under each variable that needs masking (and for later use in the diagnostic) and does not require the user to list the fx-variables as vars in the recipe; I think this is elegant and hassle-free and keeps things tidy (after all, those are auxiliary variables and should not be treated as regular variables). @bouweandela I also like the idea of, in the case masking is needed in the preprocessor,  just adding the fx-variables in the diagnostic's `variables` would allow for that to be done (note that either tasking priority with ancestors or some other priority assignment still has to be done here since the masking has to be done with already preprocessed fx variables ie cmor checked and fixed; there is also the question what sort of extra preprocessing would be needed for fx-variables - if they are used only for masking, default will be enough, but how do you split the cases if one actually needs more preprocessing done on fx-variables?) - BUT what if there is no masking  in the preprocessor and only one variable needs fx files for diagnostic purposes? We'd do the fx files retrieval and preprocessing for each regular variable and tht's redundant. This is why I keep sticking to the current implementation - it delineates two separate cases. Of course, I am very open to ideas :beer:

mattiarighi commented:
> for diagnostic use only is simple and if the user puts the needed fx variables in the recipe (like all other variables) them files will be avialable at the end of preprocessor;

This means as actual files in the `preproc_dir` and not only the path to them in `metadata.yml` / `<var>_info.ncl` as before. Correct?

valeriupredoi commented:
> > for diagnostic use only is simple and if the user puts the needed fx variables in the recipe (like all other variables) them files will be avialable at the end of preprocessor;
> 
> This means as actual files in the `preproc_dir` and not only the path to them in `metadata.yml` / `<var>_info.ncl` as before. Correct?

yes

bouweandela commented:
> for masking done in the preprocessor, ...

@valeriupredoi I would recommend leaving applying the cmor fixes to fx variables that are used as input for preprocessor functions for a later pull request, as I also wrote here: https://github.com/ESMValGroup/ESMValTool/pull/1037#issuecomment-491213895. One change per PR please.

valeriupredoi commented:
ok guys, all fx-proprietary file finders gone (now all's done with standard variable file I/O), it works well for me, not sure why the tests are failing locally, I'll have a look

valeriupredoi commented:
in any case, I will disentangle the bit that deals with fx variables cmorization and create a PR only with the fx variables correct retrieval (no cmor checks); I will, however, keep this PR active because the implementation here works well. Good weekend, chaps, off to pub :beer:

valeriupredoi commented:
OK guys, this is finally done and tested and I swear it didn't take 2 weeks :grin:

bouweandela commented:
Please re-open this pull request in the [ESMValTool Core](https://github.com/esmvalgroup/esmvalcore) repository. You can do so by pushing this branch to that repository, using the following instructions.

First add the repository with
```bash
git remote add esmvalcore git@github.com:esmvalgroup/esmvalcore
```
or 
```bash
git remote add esmvalcore https://github.com/esmvalgroup/esmvalcore
```
if you're not using ssh to connect to GitHub.

Next upload the branch with
```
git push esmvalcore your_branch
```
and open a pull request [here](https://github.com/esmvalgroup/esmvalcore/pulls)

valeriupredoi commented:
ported to https://github.com/ESMValGroup/ESMValCore/pull/19
